### ISSUE TITLE
Repeated runs (inference and/or optimisation)

### ISSUE THREAD

**Issue by MichaelClerx** [pints-team/pints#185]
@ben18785 and I are looking at repeated runs, for example to obtain multiple markov chains from adaptive MCMC (to test convergence).

We were thinking of making this part of the base class MCMC (and Optimiser) objects, like some MCMC libraries do already.

Question: Should we make this the default, e.g.:

    mcmc = SomeMCMC()
    mcmc.run(chains=4, parallel=True)

Or with a separate method:

    mcmc.repeated_run(4, parallel=True)

?

**Comment by mirams** [comment]
It isn't something everybody does, even if it should be, so might be confusing and make this look like parallel tempering or something if it is 

    mcmc.run(chains=4, parallel=True)

so I would probably go for the second...

Especially as they might want launching with different start points etc.?

**Comment by martinjrobins** [comment]
Is running multiple chains not standard? My only exposure is to STAN, and it uses option 1, so I assumed this was standard.

**Comment by ben18785** [comment]
The most popular MCMC packages all run multiple chains it as standard
(BUGS, JAGS, PYMC3 and Stan). I think that there is a lot more to be
potentially lost (as in mistakenly thinking you have converged on the
posterior when, in fact, you haven't) by not running multiple chains as
standard. When I first started doing MCMC (the likely position of many of
PINTS' potential users), I definitely made this mistake! I really strongly
think we should go for the first option.

**Comment by ben18785** [comment]
I think it's important to ask what's worse -- a little bit of confusion or a false view of convergence?

**Comment by ben18785** [comment]
If we're really clear in our examples and documentation, I wouldn't have thought that people would mistakenly think that parallel=True is parallel tempering. If we were concerned about this, can't we just change it to multiple_chains/cores=True?

Re: start points, we should, by default, start chains at disparate places in parameter space. Of course, we should allow a user to manually select a start point (e.g. a mode found by xNES etc.), if they wish to.

**Comment by mirams** [comment]
Sounds fine to me, as long as it is clear!

**Comment by MichaelClerx** [comment]
I'm leaning towards the second option myself! I find it confusing when I ask a method to run and it runs five times... Also wan't to make this decision the same for optimisation and inference methods, to keep a highly similar interface. I've never seen it in an optimisation method (except when hidden in an ipop structure).

I'm also worried about how much we'll end up cramming into the `run()` method. Would it not be better to have a super simple `run()` method and let the user build something around that? Something like

    chains = []
    for x0 in starting_points:
        chains.append(mcmc.run(x0)

is way clearer to me than a call to a single method that can also do 10 other things

**Comment by martinjrobins** [comment]
I think option 1 is the way to go. IMO running in multiple chains should be the default for the samplers. It seems to me kinda pointless running only one chain, as you have no idea if its converged.

**Comment by mirams** [comment]
I wouldn't say it's quite so clear cut. If you aren't doing anything else first, I'd agree. But if 
1) you've looked at sensitivity to all params in your experiment at your initial guess and final destination.
2) you've had a look at the likelihood bumpiness by specifically looking for local minima/other modes with an optimiser (surely a more efficient way of doing this task than running lots of mcmc chains). [Careful defined as at least running a global optimiser from all the places you'd start MCMC chains!]

Then I think you would be fine running one adaptive MCMC chain if those 2 things suggest high sensitivity and one minimum sufficently more likely than the rest is consistently recovered from optimiser. 

If you know nothing, then lots of chains, yeah.

**Comment by mirams** [comment]
In fact, in my example you can probably just do maximum likelihood and fisher information! So I suppose MCMC should sit in the I can't make too many guarantees space...

**Comment by MichaelClerx** [comment]
The question is not whether or not we should run multiple chains, the question is whether we should pack all of our cleverness into a single method call.

1. Running multiple chains, even in parallel, is 4 lines of absolute beginner code, the kind you learn in Python lesson 1. Doing it inside a method means you have to add documentation that explains this, and people need to read this and infer those 4 lines are in there, instead of just seeing them on their screen.
2. Just as packing a multiple runs into `run()` can be convenient, we'll want to start packing in schemes where you adapt e.g. the covariance  between runs. Alternatively these could go in different `run_x()` variations, but conceptually I don't think they're part of the routine at all, just something you do with those routines. I've never seen a pseudo-code mcmc that runs 5 times.
3. *This is the most important one for me*: Sticking repeated runs inside any method on the MCMC/Optimiser classes means you pass control to a single monolithic method for a very long time (days). This is undesirable for lots of reasons. You can't communicate with it, windows and/or linux will start asking if they should shut it down. You don't get debug feedback or a chance to store on disk. It's for just these reasons that I want to go in the exact opposite direction, doing away with even the single `run()` in favour of a method that does one or more iterations and then passes control back to the caller (even if only for a microsecond). This means all storage, debug output, stopping criterion testing etc. happen _outside_ of the mcmc/optimisation methods, and don't need to be implemented more than once. It also means you can implement progress bars etc. if you're using pints from another application, e.g. the web lab. 
4. If we want to do neat things like parallelising on GPUs, we'd need to pass all this information through our magic `run()` method too. How?

So as I've written in the issues about debug output, storage, and result objects. I strongly believe all our methods should work like this:

* We set up some object (MCMC class, Optimiser class)
* We set up a loop, from wherein
  * we call an iterate() method, this advances us one step and then asks us to evaluate the model on some new points
  * we do logging, debug output, talk to calling processes, check stopping criteria
  * we evaluate the model (possibly parallelised on a CPU/GPU) and pass back the information

Its then trivial to add another loop around this to repeat the process multiple times (possibly in parallel).

I'm not just making this up either, I spent a week or so converting all of Myokit's stuff to such a scheme because I had issues losing control when calling it from other threads/processes etc. (and because I wanted to get a progress bar in).
Looking back, CVODE works this way too. And CMA-ES when you use the ask-and-tell interface.

We can always add some convenience wrapper later, but I strongly believe this is how our methods should work at their core

**Comment by MichaelClerx** [comment]
At the moment, the simplest way to do multiple runs is:

    def run(index):
        np.random.seed(x)
        return mcmc.run()
    chains = pints.evaluate(run, [1,2,3,4], parallel=True)

but we should probably add a method to set the initial point, so that it becomes

chains = pints.evaluate(mcmc.run, starting_points, parallel=True)

**Comment by ben18785** [comment]
In my opinion multiple chains should definitely be the default behaviour in run.

1. Most users won't do maximum likelihood first. This is definitely suboptimal- it pays to know about your problem before starting MCMC - but reflects how people often behave. In this context it is worth doing MCMC with multiple chains.

2. Even if you do maximum likelihood estimation first, you often don't know about whether there is multimodality and/or the width of uncertainty intervals. Without multiple chains, how do you have an idea as to whether you've reached convergence?

3. Users expect this functionality as a default behaviour from other software. I think you've got to remember that most people are beginners in Python (myself included!). It's easy to forget what people won't know to do. I didn't know how to parallelise the loop. Others also won't.

**Comment by mirams** [comment]
Can we make a very lightweight wrapper which does parallel runs, which internally does what Michael suggests so that people can then delve into and copy and paste if they want to do more advanced things?

**Comment by MichaelClerx** [comment]
Agreed, we can and should add a very lightweight wrapper -- eventually -- that can do these things.

@ben18785 About your third point: Lots of our users won't have expectations from other MCMC packages (like me!). I think it's fair to say there are more potential users with Python skills than with MCMC skills. Also it's much better for people to learn some transferable python skills than to memorise how our magic methods work. In addition, parallelising is not a beginners task, we should definitely add examples and maybe even convenience methods (like we already have, really, that `evaluate` method does lots of stuff under the hood), but they shouldn't be the main way to run our stuff.

Consider running multiple parallel MCMC routines in parallel, then the one `run()` method would have to pass all the information (in both directions) to set up 2 parallelisation schemes, choose number of cores, handle debug output, etc. That's just not a scalable solution.

**Comment by ben18785** [comment]
I think you're right Michael -- perhaps people will have more Python than MCMC knowledge. But to me that's precisely why we should nudge them in the right direction - to use multiple chains as a default.

**Comment by MichaelClerx** [comment]
Fair enough, let's make sure we have some kind of convenience method and/or use multiple chains in all our examples

**Comment by ben18785** [comment]
How would this work with the MCMC result object though? Surely we want to avoid calling this on a result in a notebook?

**Comment by MichaelClerx** [comment]
> How would this work with the MCMC result object though?
I'm trying to work that out this afternoon :-)

> Surely we want to avoid calling this on a result in a notebook?
Not 100% sure what you mean by this?

**Comment by ben18785** [comment]
I mean, surely we want to have mcmc.run return an MCMCResultObject? To instantiate this object (calculate R hat, effective sample size etc.) it requires multiple chains. I don't want us having to do in a notebook:

parallel for loop: 
 Chains[i] = mcmc.run()
End

result = MCMCResultObject(chains)

Perhaps I'm not understanding, what do you mean by a convenience function?

**Comment by MichaelClerx** [comment]
Hmmmm... What I'm trying to say is that we should start from a very flexible 'hardcore' interface, and then later (e.g. a week from now, not 2019 :-)) add a convenient method somewhere that runs an MCMC routine with all the default settings. Most people will then use this, until they want more control, so this'd be the place to hint at best practices.

Sorry if I've not been to clear above, I'm just trying to work out how to design all this stuff so that it stays flexible, scalable and easy to use... It's possible we can't do all three, and have to go for flexible+scalable, and then add the "easy to use" in separate functions that use the flexible+scalable interface internally, but hide lots of their options (by design: to simplify!)

**Comment by MichaelClerx** [comment]
I'm not sure if those simplified / convenient functions would be methods of the MCMC base objects, or if they'd be separate (like the `pints.cmaes()` method we have now, for example)

**Comment by martinjrobins** [comment]
I think the idea of both a `hardcore` interface, and a convenience wrapper that runs with default settings (i.e. multiple chains) is a great idea. You mentioned CMA-ES's ask-and-tell interface, which I think works really well. I started out using CMA-ES's convenience wrapper function, then when I wanted to do something more tricky (run in parallel) I moved to the ask-and-tell interface. I'd be happy if you wanted to design something similar for pints.
